### ISSUE TITLE
Accessibility-Ready Review and Checklist

### ISSUE THREAD

**Issue by davidakennedy** [WordPress/twentynineteen#41]
Nice work overall on the accessibility of this design. My favorite parts are the nice focus states on form fields.

Here are the [requirements I tested against](https://make.wordpress.org/themes/handbook/review/accessibility/required/).

The issues below are totally fixable with a little design love. I may have missed something, so if I did, let me know in the comments. Also, please let me know if I can retest anything after PRs come in.

Anything in marked required will need to be fixed before the theme can launch. Anything marked recommended is optional for you to consider and/or fix.

**Keyboard Navigation**

Required: The focus styles look great throughout the design. There are just a few spots that need polish.

- [x] The comment `<li>`, `.comment:hover`, has a `:hover` style. However, because it's an `<li>` and not something that can receive focus, like a link, form field or button, a user can ever make the edit link appear on focus. You should be able to show the same edit link if certain links are focused on.
- [ ] Set focus styles for the audio embed block: https://cloudup.com/caSR44Mg2IA
- [x] The focus styles on posts navigation are not showing up properly: https://cloudup.com/cdp57bMMqiP

**Controls**

Pass.

**Skip Links**

Pass.

**Forms**

Required: The comment form is hidden with `display: none;`.

- [x] Forms with more than one field need to have all labels displayed visually.

**Headings**

Pass.

- [x] Recommended: Some views, like regular pages and archive pages, have more than one `<h1>`. This is okay, but not ideal because screen reader users like the clarity of one heading at level one. See _s for the ideal way to do this: https://github.com/Automattic/_s/blob/master/header.php#L31 Already mentioned in #5.

**ARIA Landmark Roles**

Pass.

- [x] Recommended: You don't need the explicit roles on the main navigation, ` role="navigation" aria-label="<?php esc_attr_e( 'Top Menu', 'twentynineteen' ); ?>` or on the `<main>` element. They can be removed.

**Content Links**

- [x] Required: Links within content must be underlined. When links appear within a larger body of block-level content, they must be clearly distinguishable from surrounding content. This would be for links in the main post body.

**Repetitive Link Text**

Pass.

**Contrasts**

Required: There are a number of colors that aren't high contrast enough to meet the requirements. Most of the colors in the design are though.

Color issues:

- [x] #919191 on #ffffff – Found in post and page meta. On search pages.
- [x] #00a0d2 on #ffffff – The hover color found on some links.

Hover colors also need to pass contrast rules.

**Images**

Pass.

**Note**: With the functionality for commenting avatars, those have a blank alt attribute, which I think is fine because those images are more decorative than anything. Some others may disagree, so I'm explaining my reasoning.

**Media**

Pass.

**Screen Reader Text**

Pass.

**Not Allowed**

* Any positive tabindex attribute: Pass.
* The inclusion of the accesskey attribute: Pass.
* Spawning new windows or tabs without warning the user: Pass.

**Comment by melchoyce** [comment]
If it helps, `#767676` is the lightest a grey can be on white before it stops passing AA.

**Comment by carolinan** [comment]
_Featured Images: The alt attribute for featured images is defined in the media manager. In any case where a featured image is displayed, the alt attribute must be included with the image._

In function twentynineteen_post_thumbnail, the user set alt text is replaced with `the_title_attribute.`
It must respect the user setting.

**Comment by carolinan** [comment]
Also,  a question about the contrast; there are still some blocks and parts of blocks in Gutenberg where the text color has a contrast that is too low. Should the theme include a fix for this? Or do we wait and hope it is fixed in Gutenberg in time.

**Comment by kjellr** [comment]
> Also, a question about the contrast; there are still some blocks and parts of blocks in Gutenberg where the text color has a contrast that is too low. Should the theme include a fix for this? Or do we wait and hope it is fixed in Gutenberg in time.

I think it depends — if it's on the front end, we should absolutely fix it. For issues within the editor, we'll need to evaluate on a case-by-case basis. Some things may be fixed by Gutenberg, but many issues can (and should) be solved by our editor styles. In general, if you see these issues, please go ahead and file bugs for them. 👍

**Comment by 2ndkauboy** [comment]
@davidakennedy I could not reproduce the first "Keyboard Navigation" issue, as I don't see any `.hover` state on `.comment` anymore.

It also seems that there is no possibility to change the focus style of the browsers `<audio>` tags in all browers. Firefox even lacks proper support of focussing the individual controls alltogether :/

Are there any particular issues you see with the focus on media tags? I could see a focus outline around the tags in Chrome, Firefox, IE 11 and Edge (IE 11 and Edge both actually lets you focus every individual control, at least on Windows 10).

**Comment by 2ndkauboy** [comment]
@davidakennedy I could not reproduce the first "Keyboard Navigation" issue, as I don't see any `.hover` state on `.comment` anymore.

It also seems that there is no possibility to change the focus style of the browsers `<audio>` tags in all browers. Firefox even lacks proper support of focussing the individual controls alltogether :/

Are there any particular issues you see with the focus on media tags? I could see a focus outline around the tags in Chrome, Firefox, IE 11 and Edge (IE 11 and Edge both actually lets you focus every individual control, at least on Windows 10).

**Comment by 2ndkauboy** [comment]
@davidakennedy I've also checked the other two open issues. It looks like the `role` attribute on the top menu was removed and I also couldn't find links in the content (and comments) without an underline style.

**Comment by kjellr** [comment]
> It looks like the role attribute on the top menu was removed and I also couldn't find links in the content (and comments) without an underline style.

Thanks, @2ndkauboy — I can confirm these and have marked them off on the list.

**Comment by kjellr** [comment]
> > The comment <li>, .comment:hover, has a :hover style. However, because it's an <li> and not something that can receive focus, like a link, form field or button, a user can ever make the edit 
> 
>  I could not reproduce the first "Keyboard Navigation" issue, as I don't see any .hover state on .comment anymore.

That's right, this was addressed in #466.

> The focus styles on posts navigation are not showing up properly: https://cloudup.com/cdp57bMMqiP

I believe this was addressed in #321: 

![screen shot 2018-11-13 at 4 10 06 pm](https://user-images.githubusercontent.com/1202812/48443319-c00fc080-e75e-11e8-82a0-c7e656c41dce.png)

I've crossed both of those off the list as well. I think the only remaining item is the one about the audio block.

**Comment by 2ndkauboy** [comment]
Unfortunately I don't understand the issue with the audio tag. The possibility to focus individual controls is very different from browser to browser (or OS). But im general, the audio tag can be focused and the play action can be triggered.

**Comment by kjellr** [comment]
> Unfortunately I don't understand the issue with the audio tag. The possibility to focus individual controls is very different from browser to browser (or OS). But im general, the audio tag can be focused and the play action can be triggered.

That's my thought too — I'm able to tab through and play/pause the controller in Chrome: 

![audio](https://user-images.githubusercontent.com/1202812/48500417-ceb2b200-e808-11e8-833e-b1e4479072a9.gif)

... but I'm unable to do so in Firefox for instance. Since this relies on the native browser `audio` component, I'm not sure there's anything we can do about that.

**Comment by allancole** [comment]
> Set focus styles for the audio embed block: https://cloudup.com/caSR44Mg2IA

Since this appears to be an issue with how Browsers deal with the `audio` tag, the problem is not a caused by the theme. 

Closing this since all other items have been addressed :-)
